File: sdk/sdk-py/glimpse_sdk/protocol.py
from __future__ import annotations

from dataclasses import dataclass, field, replace


@dataclass(slots=True)
class StatusItem:
    id: str | None = None
    icon: str | None = None
    label: str | None = None
    tooltip: str | None = None
    css_classes: list[str] = field(default_factory=list)

    def with_css_class(self, css_class: str) -> "StatusItem":
        """Return a copy with ``css_class`` appended. Immutable-update form
        mirroring the Rust SDK's ``css_class()`` builder so SDK-hopping users
        see equivalent ergonomics."""
        return replace(self, css_classes=[*self.css_classes, css_class])

    def to_protocol(self) -> dict[str, object]:
        payload: dict[str, object] = {}
        if self.id is not None:
            payload["id"] = self.id
        if self.icon is not None:
            payload["icon"] = self.icon
        if self.label is not None:
            payload["label"] = self.label
        if self.tooltip is not None:
            payload["tooltip"] = self.tooltip
        if self.css_classes:
            payload["css_classes"] = list(self.css_classes)
        return payload

File: sdk/sdk-ts/src/protocol.ts
export class StatusItem {
  constructor(
    public readonly options: {
      id?: string;
      icon?: string;
      label?: string;
      tooltip?: string;
      cssClasses?: string[];
    } = {},
  ) {}

  /**
   * Return a copy with `cssClass` appended to `cssClasses`. Immutable-update
   * form that mirrors the Rust SDK's `css_class()` builder so SDK-hopping
   * users see equivalent ergonomics across languages.
   */
  withCssClass(cssClass: string): StatusItem {
    return new StatusItem({
      ...this.options,
      cssClasses: [...(this.options.cssClasses ?? []), cssClass],
    });
  }

  toProtocol(): Record<string, unknown> {
    const payload: Record<string, unknown> = {};
    if (this.options.id !== undefined) {
      payload.id = this.options.id;
    }
    if (this.options.icon !== undefined) {
      payload.icon = this.options.icon;
    }
    if (this.options.label !== undefined) {
      payload.label = this.options.label;
    }
    if (this.options.tooltip !== undefined) {
      payload.tooltip = this.options.tooltip;
    }
    if (this.options.cssClasses && this.options.cssClasses.length > 0) {
      payload.css_classes = [...this.options.cssClasses];
    }
    return payload;
  }
}
